shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -37,17 +37,16 @@ class JSBackend {
     // should returns ("{ x, y }", ["x", "y"])
     case Rcd(fields) => {
       // $ means an insertion point
-      fields.foldLeft[Str -> Ls[Str]]("{ $ }" -> List.empty) {
-        case ((code, names), (key, value)) =>
-          value match {
-            case Var(_) => code.replace("$", s"$key, ") -> (names :+ key)
-            case rcd: Rcd => {
-              val (subCode, subNames) = translateLetPattern(rcd)
-              code.replace("$", s"$key: ${subCode}, ") -> (names ++ subNames)
-            }
-            // Is there any other cases?
-            case _ => ???
+      fields.foldLeft[Str -> Ls[Str]]("{ $ }" -> List.empty) { case ((code, names), (key, value)) =>
+        value match {
+          case Var(_) => code.replace("$", s"$key, ") -> (names :+ key)
+          case rcd: Rcd => {
+            val (subCode, subNames) = translateLetPattern(rcd)
+            code.replace("$", s"$key: ${subCode}, ") -> (names ++ subNames)
           }
+          // Is there any other cases?
+          case _ => ???
+        }
       }
```

Comment:
Ugh, I just noticed this part with the `$` substitutions. That's _very_ funky logic, let's not do that!! There are much cleaner ways of building this string. BTW, `$` is a valid character in JS and Scala identifier names... Also, let's replace the YOLO `case _ => ???` with an enumeration of all terms form that can be here and a proper error reading "unsupported pattern shape" 🙂

Reply:
Hi, I replaced the string replacement logic with AST construction.

Reply:
@chengluyu don't forget to mark as resolved when it is. I think this is resolved, right?